prow/plugins/projectqueries/projectqueries.go

```diff
@@ -0,0 +1,208 @@
+)
+
+// TODO Create a new handler for issues, look in hook/server.go
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The projects plugin auto adds issues/PRs to projects based on a search query",
+		Config: func(config *plugins.Configuration) map[string]string {
+			configMap := make(map[string]string)
+			configString := ""
+			for _, projectQuery := range config.ProjectQueries {
+				projectString := fmt.Sprintf("Project Org: \"%s\", Project Name: \"%s\", Project Column: \"%s\"", projectQuery.Org, projectQuery.Name, projectQuery.Column)
+				labelsString := ""
+				for _, label := range projectQuery.Query.Labels {
+					labelsString = fmt.Sprintf("%s label:%s", labelsString, label)
+				}
+				queryString := fmt.Sprintf("Query: \"org:%s state:%s%s\"", projectQuery.Query.Org, projectQuery.Query.State, labelsString)
+				configString = fmt.Sprintf("%s\n%s %s", configString, projectString, queryString)
+			}
+			configMap[""] = configString
+			return configMap
+		}(config),
+	}
+	return pluginHelp, nil
+}
+
+func handlePullRequest(pc plugins.Agent, pe github.PullRequestEvent) error {
+	return handlePR(pc.GitHubClient, pc.PluginConfig.ProjectQueries, pc.Logger, pe)
+}
+
+// Strict subset of *github.Client methods.
+type githubClient interface {
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	GetOrgProjects(org string) ([]github.Project, error)
+	GetProjectColumns(projectID int) ([]github.ProjectColumn, error)
+	CreateProjectCard(columnID int, projectCard github.ProjectCard) (*github.ProjectCard, error)
+}
+
+func handlePR(gc githubClient, projectQueries []plugins.ProjectQuery, le *logrus.Entry, pe github.PullRequestEvent) error {
+
+	if len(projectQueries) == 0 {
+		return nil
+	}
+
+	// Only respond to label add or issue/PR open events
+	if !isPRChanged(pe) {
+		return nil
+	}
+
+	// Get any ProjectQueries that match this PR
+	matchedProjectQueries, err := getMatchingProjectQueries(gc, projectQueries, pe)
+	if err != nil {
+		return err
+	}
+	if len(matchedProjectQueries) == 0 {
+		return nil
+	}
+
+	for _, matchedProjectQuery := range matchedProjectQueries {
+		err = addPRToProject(gc, matchedProjectQuery, pe)
+		le.WithError(err)
+	}
+
+	return nil
+}
+
+// labelsInLabels check that all of the queryLabels (the first list), are present in
+// the list of github.Labels (the second list)
+func labelsInLabels(queryLabels []string, labels []github.Label) bool {
+	for _, queryLabel := range queryLabels {
+		match := false
+		for _, label := range labels {
+			if queryLabel == label.Name {
+				match = true
+				break
+			}
+		}
+		if match == false {
+			return false
+		}
+	}
+	return true
+}
+
+// These are the only actions that this plugin cares about
+func isPRChanged(pe github.PullRequestEvent) bool {
```

Comment:
I'm not sure this method brings much value over:

```go
if pe.Action != github.PullRequestActionOpened {
    return nil
}
```

At the callsite

Reply:
Wouldn't this miss pull requests that had relevant labels added after opening?

Reply:
I guess the larger point was -- we can enumerate the list of actions in the `if` statement as done in the other plugins, not much benefit to break out into a separate method and explicitly deny other actions

Reply:
I've now changed this to use an if statement and removed the `isPRChanged` function.